### PR TITLE
[flutter web] Listen for service extension registration events to determine hot-restart method name

### PR DESCRIPTION
**Pull request by elliette**
### Some background:

* Flutter registers the hot-restart service extension for all devices **except** web devices. For web devices, DWDS registers the service extensions.
* When a user is debugging their code via VS Code, the VS Code Dart extension [sends a hot-restart](https://github.com/Dart-Code/Dart-Code/blob/94cb81c5526adc7c8a1203cba201e8c1d99adf82/src/debug/run_daemon_base.ts#L100) request via stdin/out to flutter_tools
* flutter_tools then [calls the "hotRestart" service extension](https://github.com/flutter/flutter/blob/f3978c7a46f083c8e91690b8b8be3e03b1e38155/packages/flutter_tools/lib/src/isolated/resident_web_runner.dart#L447) (which, again, has been registered by DWDS)

### Why is this change necessary?

In DWDS, we are changing how we register the "hotRestart" service extension (here is the PR for that: https://github.com/dart-lang/webdev/pull/2388). Previously, we registered the "hotRestart" service extension on a client that was directly connected to the VmService. With https://github.com/dart-lang/webdev/pull/2388, we will be registering the "hotRestart" service extension on a client that is connected to DDS. 

When a service extension is registered against DDS, DDS adds a prefix to the service extension method name (e.g. "hotRestart" becomes "s0.hotRestart"). It informs clients of the service extension name via `kServiceRegistered` events sent to the `Service` stream. 

Therefore, this change simply listens to those service extension registered events, and uses them to determine the "hotRestart" service extension's method name.

### CODE REVIEW

**Review by DanTup — APPROVED** [review]
The seems reasonable to me - we have to [do the same in Dart-Code for calling services](https://github.com/Dart-Code/Dart-Code/blob/94cb81c5526adc7c8a1203cba201e8c1d99adf82/src/extension/flutter/vm_service_extensions.ts#L249). I never understood why the names might be different.. but Jacob recently mentioned it might be because multiple clients can register the same service names, and this is how they'd be disambiguated. Dart-Code just calls the last one with that name, so if there are multiple they might not all be being called!

(I don't know the best way to add tests for this though, I'll defer to Chris for suggestions)

**Comment by christopherfujino** [comment]
I was gonna say, "I'm assuming this is covered by existing tests", but then saw the test failures :) This LGTM once you update the test expectations:

```
00:05 +156 -18: test/general.shard/resident_web_runner_test.dart: cleanup of resources is safe to call multiple times [E]
  Expected: <Instance of 'Map<String, Object?>'> with `method`: 'streamListen' and `args`: {'streamId': 'Isolate'}
    Actual: {
              'jsonrpc': '2.0',
              'id': '53',
              'method': 'streamListen',
              'params': {'streamId': 'Service'}
            }
```

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**
